Module-CoreList-2.17

diff -urN perl-5.10.0.orig/lib/Module/CoreList/bin/corelist perl-5.10.0/lib/Module/CoreList/bin/corelist
--- perl-5.10.0.orig/lib/Module/CoreList/bin/corelist	2007-12-18 11:47:07.000000000 +0100
+++ perl-5.10.0/lib/Module/CoreList/bin/corelist	2009-02-10 11:51:46.000000000 +0100
@@ -11,14 +11,14 @@
 =head1 SYNOPSIS
 
     corelist -v
-    corelist [-a] <ModuleName> | /<ModuleRegex>/ [<ModuleVersion>] ...
+    corelist [-a|-d] <ModuleName> | /<ModuleRegex>/ [<ModuleVersion>] ...
     corelist [-v <PerlVersion>] [ <ModuleName> | /<ModuleRegex>/ ] ...
 
 =head1 OPTIONS
 
 =over
 
-=item -a modulename
+=item -a
 
 lists all versions of the given module (or the matching modules, in case you
 used a module regexp) in the perls Module::CoreList knows about.
@@ -44,6 +44,11 @@
       5.009002   1.04
       5.009003   1.06
 
+=item -d
+
+finds the first perl version where a module has been released by
+date, and not by version number (as is the default).
+
 =item -? or -help
 
 help! help! help! to see more help, try --man.
@@ -79,7 +84,7 @@
 
 my %Opts;
 
-GetOptions(\%Opts, qw[ help|?! man! v|version:f a! ] );
+GetOptions(\%Opts, qw[ help|?! man! v|version:f a! d ] );
 
 pod2usage(1) if $Opts{help};
 pod2usage(-verbose=>2) if $Opts{man};
@@ -93,15 +98,16 @@
     }
 
     $Opts{v} = numify_version( $Opts{v} );
-    if( !exists $Module::CoreList::version{$Opts{v}} ) {
+    my $version_hash = Module::CoreList->find_version($Opts{v});
+    if( !$version_hash ) {
         print "\nModule::CoreList has no info on perl v$Opts{v}\n\n";
         exit 1;
     }
 
     if ( !@ARGV ) {
 	print "\nThe following modules were in perl v$Opts{v} CORE\n";
-	print "$_ ", $Module::CoreList::version{$Opts{v}}{$_} || " ","\n"
-	for sort keys %{$Module::CoreList::version{$Opts{v}}};
+	print "$_ ", $version_hash->{$_} || " ","\n"
+	for sort keys %$version_hash;
 	print "\n";
 	exit 0;
     }
@@ -149,12 +155,17 @@
     my($mod,$ver) = @_;
 
     if ( $Opts{v} ) {
-        return printf "  %-24s %-10s\n",
-            $mod,
-            $Module::CoreList::version{$Opts{v}}{$mod} || 'undef';
+	my $version_hash = Module::CoreList->find_version($Opts{v});
+	if ($version_hash) {
+	    print $mod, " ", $version_hash->{$mod} || 'undef', "\n";
+	    return;
+	}
+	else { die "Shouldn't happen" }
     }
 
-    my $ret = Module::CoreList->first_release(@_);
+    my $ret = $Opts{d}
+	? Module::CoreList->first_release_by_date(@_)
+	: Module::CoreList->first_release(@_);
     my $msg = $mod;
     $msg .= " $ver" if $ver;
 
@@ -184,13 +195,12 @@
 
 sub numify_version {
     my $ver = shift;
-    if ( index( $ver, q{.}, index( $ver, q{.} ) ) >= 0 ) {
-	eval { require version };
-	if ($@) {
-	    die "You need to install version.pm to use dotted version numbers\n";
-	}
+    if ($ver =~ /\..+\./) {
+	eval { require version ; 1 }
+	    or die "You need to install version.pm to use dotted version numbers\n";
         $ver = version->new($ver)->numify;
     }
+    $ver += 0;
     return $ver;
 }
 
diff -urN perl-5.10.0.orig/lib/Module/CoreList.pm perl-5.10.0/lib/Module/CoreList.pm
--- perl-5.10.0.orig/lib/Module/CoreList.pm	2007-12-18 11:47:07.000000000 +0100
+++ perl-5.10.0/lib/Module/CoreList.pm	2009-02-10 11:51:14.000000000 +0100
@@ -1,7 +1,7 @@
 package Module::CoreList;
 use strict;
 use vars qw/$VERSION %released %patchlevel %version %families/;
-$VERSION = '2.13';
+$VERSION = '2.17';
 
 =head1 NAME
 
@@ -59,7 +59,7 @@
 
 Module::CoreList currently covers the 5.000, 5.001, 5.002, 5.003_07, 5.004,
 5.004_05, 5.005, 5.005_03, 5.005_04, 5.6.0, 5.6.1, 5.6.2, 5.7.3, 5.8.0, 5.8.1,
-5.8.2, 5.8.3, 5.8.4, 5.8.5, 5.8.6, 5.8.7, 5.8.8, 5.9.0, 5.9.1, 5.9.2, 5.9.3,
+5.8.2, 5.8.3, 5.8.4, 5.8.5, 5.8.6, 5.8.7, 5.8.8, 5.8.9, 5.9.0, 5.9.1, 5.9.2, 5.9.3,
 5.9.4, 5.9.5 and 5.10.0 releases of perl.
 
 =head1 HISTORY
@@ -74,7 +74,7 @@
 
 =head1 COPYRIGHT
 
-Copyright (C) 2002-2007 Richard Clamp.  All Rights Reserved.
+Copyright (C) 2002-2009 Richard Clamp.  All Rights Reserved.
 
 This module is free software; you can redistribute it and/or modify it
 under the same terms as Perl itself.
@@ -138,12 +138,17 @@
     return sort keys %mods
 }
 
+sub find_version {
+    my ($class, $v) = @_;
+    return $version{$v} if defined $version{$v};
+    return undef;
+}
 
 # when things escaped
 %released = (
     5.000    => '1994-10-17',
     5.001    => '1995-03-14',
-    5.002    => '1996-02-96',
+    5.002    => '1996-02-29',
     5.00307  => '1996-10-10',
     5.004    => '1997-05-15',
     5.005    => '1998-07-22',
@@ -170,13 +175,14 @@
     5.009004 => '2006-08-15',
     5.009005 => '2007-07-07',
     5.010000 => '2007-12-18',
+    5.008009 => '2008-12-14',
    );
 
 # perforce branches and patch levels
 %patchlevel = (
     5.005    => [perl => 1647],
     5.00503  => ['maint-5.005' => 3198],
-    5.00405  => ['maint-5.004' => 999],
+    5.00405  => ['maint-5.004' => 3296],
     5.006    => [perl => 5899],
     5.006001 => ['maint-5.6' => 9654],
     5.006002 => ['maint-5.6' => 21727],
@@ -198,6 +204,7 @@
     5.009004 => [perl => 28727],
     5.009005 => [perl => 31562],
     5.010000 => [perl => 32642],
+    5.008009 => ['maint-5.8' => 35095],
 );
 
 for my $version ( sort { $a <=> $b } keys %released ) {
@@ -1377,6 +1384,7 @@
         'ExtUtils::MM_Win32'    => undef, #./lib/ExtUtils/MM_Win32.pm
         'ExtUtils::MakeMaker'   => '5.45', #./lib/ExtUtils/MakeMaker.pm
         'ExtUtils::Manifest'    => '1.33 ', #./lib/ExtUtils/Manifest.pm
+        'ExtUtils::Miniperl'    => undef,
         'ExtUtils::Mkbootstrap' => '1.14 ', #./lib/ExtUtils/Mkbootstrap.pm
         'ExtUtils::Mksymlists'  => '1.17 ', #./lib/ExtUtils/Mksymlists.pm
         'ExtUtils::Packlist'    => '0.03', #./lib/ExtUtils/Packlist.pm
@@ -1601,6 +1609,7 @@
         'ExtUtils::Liblist'     => 1.26 ,
         'ExtUtils::MakeMaker'   => 5.45,
         'ExtUtils::Manifest'    => 1.33 ,
+        'ExtUtils::Miniperl'    => undef,
         'ExtUtils::Mkbootstrap' => 1.14 ,
         'ExtUtils::Mksymlists'  => 1.17 ,
         'ExtUtils::MM_Cygwin'   => undef,
@@ -2002,7 +2011,7 @@
         'warnings'              => undef, #lib/warnings.pm
         'warnings::register'    => undef, #lib/warnings/register.pm
         'XSLoader'              => '0.01', #lib/XSLoader.pm
-   },
+    },
 
     5.007003   => {
         'AnyDBM_File'           => '1.00',
@@ -2103,6 +2112,7 @@
         'ExtUtils::Liblist'     => '1.2701',
         'ExtUtils::MakeMaker'   => '5.48_03',
         'ExtUtils::Manifest'    => '1.35',
+        'ExtUtils::Miniperl'    => undef,
         'ExtUtils::Mkbootstrap' => '1.1401',
         'ExtUtils::Mksymlists'  => '1.18',
         'ExtUtils::MM_BeOS'     => '1.00',
@@ -2424,6 +2434,7 @@
         'ExtUtils::Liblist::Kid'=> '1.29', #./lib/ExtUtils/Liblist/Kid.pm
         'ExtUtils::MakeMaker'   => '6.03', #./lib/ExtUtils/MakeMaker.pm
         'ExtUtils::Manifest'    => '1.38', #./lib/ExtUtils/Manifest.pm
+        'ExtUtils::Miniperl'    => undef,
         'ExtUtils::Mkbootstrap' => '1.15', #./lib/ExtUtils/Mkbootstrap.pm
         'ExtUtils::Mksymlists'  => '1.19', #./lib/ExtUtils/Mksymlists.pm
         'ExtUtils::MM'          => '0.04', #./lib/ExtUtils/MM.pm
@@ -4457,6 +4468,7 @@
         'XSLoader'              => '0.03',  #lib/XSLoader.pm
         'XS::Typemap'           => '0.01',  #lib/XS/Typemap.pm
     },
+
     5.008004 => {
         'AnyDBM_File'           => '1.00',  #lib/AnyDBM_File.pm
         'attributes'            => '0.06',  #lib/attributes.pm
@@ -5550,6 +5562,7 @@
         'XSLoader'              => '0.02',  #lib/XSLoader.pm
         'XS::Typemap'           => '0.01',  #lib/XS/Typemap.pm
     },
+
     5.009002 => {
 	'AnyDBM_File'           => '1.00',
 	'Attribute::Handlers'   => '0.78_01',
@@ -5590,6 +5603,7 @@
 	'Carp::Heavy'           => '1.04',
 	'Class::ISA'            => '0.33',
 	'Class::Struct'         => '0.63',
+        'Config'                => undef,
 	'Config::Extensions'    => '0.01',
 	'Cwd'                   => '3.05',
 	'DB'                    => '1.0',
@@ -5668,6 +5682,7 @@
 	'ExtUtils::MakeMaker::bytes'=> '0.01',
 	'ExtUtils::MakeMaker::vmsish'=> '0.01',
 	'ExtUtils::Manifest'    => '1.44',
+        'ExtUtils::Miniperl'    => undef,
 	'ExtUtils::Mkbootstrap' => '1.15',
 	'ExtUtils::Mksymlists'  => '1.19',
 	'ExtUtils::Packlist'    => '0.04',
@@ -5914,6 +5929,7 @@
 	'warnings'              => '1.04',
 	'warnings::register'    => '1.00',
     },
+
     5.008007 => {
 	'AnyDBM_File'           => '1.00',
 	'Attribute::Handlers'   => '0.78_01',
@@ -5954,6 +5970,7 @@
 	'Carp::Heavy'           => '1.04',
 	'Class::ISA'            => '0.33',
 	'Class::Struct'         => '0.63',
+        'Config'                => undef,
 	'Cwd'                   => '3.05',
 	'DB'                    => '1.0',
 	'DBM_Filter'            => '0.01',
@@ -6031,6 +6048,7 @@
 	'ExtUtils::MakeMaker::bytes'=> '0.01',
 	'ExtUtils::MakeMaker::vmsish'=> '0.01',
 	'ExtUtils::Manifest'    => '1.42',
+        'ExtUtils::Miniperl'    => undef,
 	'ExtUtils::Mkbootstrap' => '1.15',
 	'ExtUtils::Mksymlists'  => '1.19',
 	'ExtUtils::Packlist'    => '0.04',
@@ -6273,6 +6291,7 @@
 	'warnings'              => '1.03',
 	'warnings::register'    => '1.00',
     },
+
     5.009003 => {
 	'AnyDBM_File'           => '1.00',
 	'Archive::Tar'          => '1.26_01',
@@ -6343,6 +6362,7 @@
 	'Compress::Zlib::ParseParameters'=> '2.000_07',
 	'Compress::Zlib::UncompressPlugin::Identity'=> '2.000_05',
 	'Compress::Zlib::UncompressPlugin::Inflate'=> '2.000_05',
+        'Config'                => undef,
 	'Config::Extensions'    => '0.01',
 	'Cwd'                   => '3.15',
 	'DB'                    => '1.01',
@@ -6438,6 +6458,7 @@
 	'ExtUtils::MakeMaker::bytes'=> '0.01',
 	'ExtUtils::MakeMaker::vmsish'=> '0.01',
 	'ExtUtils::Manifest'    => '1.46',
+        'ExtUtils::Miniperl'    => undef,
 	'ExtUtils::Mkbootstrap' => '1.15',
 	'ExtUtils::Mksymlists'  => '1.19',
 	'ExtUtils::Packlist'    => '0.04',
@@ -6722,6 +6743,7 @@
 	'warnings'              => '1.05',
 	'warnings::register'    => '1.01',
     },
+
     5.008008 => {
 	'AnyDBM_File'           => '1.00',
 	'Attribute::Handlers'   => '0.78_02',
@@ -6762,6 +6784,7 @@
 	'Carp::Heavy'           => '1.04',
 	'Class::ISA'            => '0.33',
 	'Class::Struct'         => '0.63',
+        'Config'                => undef,
 	'Cwd'                   => '3.12',
 	'DB'                    => '1.01',
 	'DBM_Filter'            => '0.01',
@@ -6844,6 +6867,7 @@
 	'ExtUtils::MakeMaker::bytes'=> '0.01',
 	'ExtUtils::MakeMaker::vmsish'=> '0.01',
 	'ExtUtils::Manifest'    => '1.46',
+        'ExtUtils::Miniperl'    => undef,
 	'ExtUtils::Mkbootstrap' => '1.15',
 	'ExtUtils::Mksymlists'  => '1.19',
 	'ExtUtils::Packlist'    => '0.04',
@@ -7089,6 +7113,7 @@
 	'warnings'              => '1.05',
 	'warnings::register'    => '1.01',
     },
+
     5.009004 => {
 	'AnyDBM_File'           => '1.00',
 	'Archive::Tar'          => '1.30_01',
@@ -7138,6 +7163,7 @@
 	'Class::Struct'         => '0.63',
 	'Compress::Raw::Zlib'   => '2.000_13',
 	'Compress::Zlib'        => '2.000_13',
+        'Config'                => undef,
 	'Config::Extensions'    => '0.01',
 	'Cwd'                   => '3.19',
 	'DB'                    => '1.01',
@@ -7233,6 +7259,7 @@
 	'ExtUtils::MakeMaker::bytes'=> '0.01',
 	'ExtUtils::MakeMaker::vmsish'=> '0.01',
 	'ExtUtils::Manifest'    => '1.46_01',
+        'ExtUtils::Miniperl'    => undef,
 	'ExtUtils::Mkbootstrap' => '1.15_01',
 	'ExtUtils::Mksymlists'  => '1.19_01',
 	'ExtUtils::Packlist'    => '1.41',
@@ -7571,6 +7598,7 @@
 	'warnings'              => '1.05',
 	'warnings::register'    => '1.01',
     },
+
     5.009005 => {
 	'AnyDBM_File'           => '1.00',
 	'Archive::Extract'      => '0.22_01',
@@ -8105,6 +8133,7 @@
 	'warnings'              => '1.06',
 	'warnings::register'    => '1.01',
     },
+
     5.010000 => {
 	'AnyDBM_File'           => '1.00',
 	'Archive::Extract'      => '0.24',
@@ -8285,6 +8314,7 @@
 	'ExtUtils::MakeMaker::bytes'=> '6.42',
 	'ExtUtils::MakeMaker::vmsish'=> '6.42',
 	'ExtUtils::Manifest'    => '1.51_01',
+        'ExtUtils::Miniperl'    => undef,
 	'ExtUtils::Mkbootstrap' => '6.42',
 	'ExtUtils::Mksymlists'  => '6.42',
 	'ExtUtils::Packlist'    => '1.43',
@@ -8644,7 +8674,418 @@
 	'warnings'              => '1.06',
 	'warnings::register'    => '1.01',
     },
+
+    5.008009 => {
+	'AnyDBM_File'           => '1.00',
+	'Attribute::Handlers'   => '0.78_03',
+	'AutoLoader'            => '5.67',
+	'AutoSplit'             => '1.06',
+	'B'                     => '1.19',
+	'B::Asmdata'            => '1.02',
+	'B::Assembler'          => '0.08',
+	'B::Bblock'             => '1.02_01',
+	'B::Bytecode'           => '1.01_01',
+	'B::C'                  => '1.05',
+	'B::CC'                 => '1.00_01',
+	'B::Concise'            => '0.76',
+	'B::Debug'              => '1.05',
+	'B::Deparse'            => '0.87',
+	'B::Disassembler'       => '1.05',
+	'B::Lint'               => '1.11',
+	'B::Lint::Debug'        => undef,
+	'B::Showlex'            => '1.02',
+	'B::Stackobj'           => '1.00',
+	'B::Stash'              => '1.00',
+	'B::Terse'              => '1.05',
+	'B::Xref'               => '1.01',
+	'Benchmark'             => '1.1',
+	'ByteLoader'            => '0.06',
+	'CGI'                   => '3.42',
+	'CGI::Apache'           => '1.00',
+	'CGI::Carp'             => '1.30_01',
+	'CGI::Cookie'           => '1.29',
+	'CGI::Fast'             => '1.07',
+	'CGI::Pretty'           => '1.08',
+	'CGI::Push'             => '1.04',
+	'CGI::Switch'           => '1.00',
+	'CGI::Util'             => '1.5_01',
+	'CPAN'                  => '1.9301',
+	'CPAN::Debug'           => '5.5',
+	'CPAN::DeferedCode'     => '5.50',
+	'CPAN::Distroprefs'     => '6',
+	'CPAN::FirstTime'       => '5.5_01',
+	'CPAN::HandleConfig'    => '5.5',
+	'CPAN::Kwalify'         => '5.50',
+	'CPAN::Nox'             => '5.50',
+	'CPAN::Queue'           => '5.5',
+	'CPAN::Tarzip'          => '5.5',
+	'CPAN::Version'         => '5.5',
+	'Carp'                  => '1.10',
+	'Carp::Heavy'           => '1.10',
+	'Class::ISA'            => '0.33',
+	'Class::Struct'         => '0.63',
+	'Config'                => undef,
+	'Cwd'                   => '3.29',
+	'DB'                    => '1.01',
+	'DBM_Filter'            => '0.02',
+	'DBM_Filter::compress'  => '0.02',
+	'DBM_Filter::encode'    => '0.02',
+	'DBM_Filter::int32'     => '0.02',
+	'DBM_Filter::null'      => '0.02',
+	'DBM_Filter::utf8'      => '0.02',
+	'DB_File'               => '1.817',
+	'DCLsym'                => '1.03',
+	'Data::Dumper'          => '2.121_17',
+	'Devel::DProf'          => '20080331.00',
+	'Devel::InnerPackage'   => '0.3',
+	'Devel::PPPort'         => '3.14',
+	'Devel::Peek'           => '1.04',
+	'Devel::SelfStubber'    => '1.03',
+	'Digest'                => '1.15',
+	'Digest::MD5'           => '2.37',
+	'Digest::base'          => '1.00',
+	'Digest::file'          => '1.00',
+	'DirHandle'             => '1.02',
+	'Dumpvalue'             => '1.12',
+	'DynaLoader'            => '1.09',
+	'Encode'                => '2.26',
+	'Encode::Alias'         => '2.10',
+	'Encode::Byte'          => '2.03',
+	'Encode::CJKConstants'  => '2.02',
+	'Encode::CN'            => '2.02',
+	'Encode::CN::HZ'        => '2.05',
+	'Encode::Config'        => '2.05',
+	'Encode::EBCDIC'        => '2.02',
+	'Encode::Encoder'       => '2.01',
+	'Encode::Encoding'      => '2.05',
+	'Encode::GSM0338'       => '2.01',
+	'Encode::Guess'         => '2.02',
+	'Encode::JP'            => '2.03',
+	'Encode::JP::H2Z'       => '2.02',
+	'Encode::JP::JIS7'      => '2.04',
+	'Encode::KR'            => '2.02',
+	'Encode::KR::2022_KR'   => '2.02',
+	'Encode::MIME::Header'  => '2.05',
+	'Encode::MIME::Header::ISO_2022_JP'=> '1.03',
+	'Encode::MIME::Name'    => '1.01',
+	'Encode::Symbol'        => '2.02',
+	'Encode::TW'            => '2.02',
+	'Encode::Unicode'       => '2.05',
+	'Encode::Unicode::UTF7' => '2.04',
+	'English'               => '1.03',
+	'Env'                   => '1.00',
+	'Errno'                 => '1.10',
+	'Exporter'              => '5.63',
+	'Exporter::Heavy'       => '5.63',
+	'ExtUtils::Command'     => '1.15',
+	'ExtUtils::Command::MM' => '6.48',
+	'ExtUtils::Constant'    => '0.21',
+	'ExtUtils::Constant::Base'=> '0.04',
+	'ExtUtils::Constant::ProxySubs'=> '0.06',
+	'ExtUtils::Constant::Utils'=> '0.02',
+	'ExtUtils::Constant::XS'=> '0.02',
+	'ExtUtils::Embed'       => '1.28',
+	'ExtUtils::Install'     => '1.50_01',
+	'ExtUtils::Installed'   => '1.43',
+	'ExtUtils::Liblist'     => '6.48',
+	'ExtUtils::Liblist::Kid'=> '6.48',
+	'ExtUtils::MM'          => '6.48',
+	'ExtUtils::MM_AIX'      => '6.48',
+	'ExtUtils::MM_Any'      => '6.48',
+	'ExtUtils::MM_BeOS'     => '6.48',
+	'ExtUtils::MM_Cygwin'   => '6.48',
+	'ExtUtils::MM_DOS'      => '6.48',
+	'ExtUtils::MM_Darwin'   => '6.48',
+	'ExtUtils::MM_MacOS'    => '6.48',
+	'ExtUtils::MM_NW5'      => '6.48',
+	'ExtUtils::MM_OS2'      => '6.48',
+	'ExtUtils::MM_QNX'      => '6.48',
+	'ExtUtils::MM_UWIN'     => '6.48',
+	'ExtUtils::MM_Unix'     => '6.48',
+	'ExtUtils::MM_VMS'      => '6.48',
+	'ExtUtils::MM_VOS'      => '6.48',
+	'ExtUtils::MM_Win32'    => '6.48',
+	'ExtUtils::MM_Win95'    => '6.48',
+	'ExtUtils::MY'          => '6.48',
+	'ExtUtils::MakeMaker'   => '6.48',
+	'ExtUtils::MakeMaker::Config'=> '6.48',
+	'ExtUtils::MakeMaker::bytes'=> '6.48',
+	'ExtUtils::MakeMaker::vmsish'=> '6.48',
+	'ExtUtils::Manifest'    => '1.55',
+	'ExtUtils::Miniperl'    => undef,
+	'ExtUtils::Mkbootstrap' => '6.48',
+	'ExtUtils::Mksymlists'  => '6.48',
+	'ExtUtils::Packlist'    => '1.43',
+	'ExtUtils::ParseXS'     => '2.19',
+	'ExtUtils::testlib'     => '6.48',
+	'Fatal'                 => '1.06',
+	'Fcntl'                 => '1.06',
+	'File::Basename'        => '2.77',
+	'File::CheckTree'       => '4.4',
+	'File::Compare'         => '1.1005',
+	'File::Copy'            => '2.13',
+	'File::DosGlob'         => '1.01',
+	'File::Find'            => '1.13',
+	'File::Glob'            => '1.06',
+	'File::Path'            => '2.07_02',
+	'File::Spec'            => '3.29',
+	'File::Spec::Cygwin'    => '3.29',
+	'File::Spec::Epoc'      => '3.29',
+	'File::Spec::Functions' => '3.29',
+	'File::Spec::Mac'       => '3.29',
+	'File::Spec::OS2'       => '3.29',
+	'File::Spec::Unix'      => '3.29',
+	'File::Spec::VMS'       => '3.29',
+	'File::Spec::Win32'     => '3.29',
+	'File::Temp'            => '0.20',
+	'File::stat'            => '1.01',
+	'FileCache'             => '1.07',
+	'FileHandle'            => '2.01',
+	'Filespec'              => '1.11',
+	'Filter::Simple'        => '0.83',
+	'Filter::Util::Call'    => '1.07',
+	'FindBin'               => '1.49',
+	'GDBM_File'             => '1.09',
+	'Getopt::Long'          => '2.37',
+	'Getopt::Std'           => '1.06',
+	'Hash::Util'            => '0.06',
+	'I18N::Collate'         => '1.00',
+	'I18N::LangTags'        => '0.35',
+	'I18N::LangTags::Detect'=> '1.03',
+	'I18N::LangTags::List'  => '0.35',
+	'I18N::Langinfo'        => '0.02',
+	'IO'                    => '1.23',
+	'IO::Dir'               => '1.06',
+	'IO::File'              => '1.14',
+	'IO::Handle'            => '1.27',
+	'IO::Pipe'              => '1.13',
+	'IO::Poll'              => '0.07',
+	'IO::Seekable'          => '1.10',
+	'IO::Select'            => '1.17',
+	'IO::Socket'            => '1.30',
+	'IO::Socket::INET'      => '1.31',
+	'IO::Socket::UNIX'      => '1.23',
+	'IPC::Msg'              => '2.00',
+	'IPC::Open2'            => '1.03',
+	'IPC::Open3'            => '1.03',
+	'IPC::Semaphore'        => '2.00',
+	'IPC::SharedMem'        => '2.00',
+	'IPC::SysV'             => '2.00',
+	'IPC::lib::IPC::Msg'    => '2.00',
+	'IPC::lib::IPC::Semaphore'=> '2.00',
+	'IPC::lib::IPC::SharedMem'=> '2.00',
+	'List::Util'            => '1.19',
+	'Locale::Constants'     => '2.07',
+	'Locale::Country'       => '2.07',
+	'Locale::Currency'      => '2.07',
+	'Locale::Language'      => '2.07',
+	'Locale::Maketext'      => '1.13',
+	'Locale::Maketext::Guts'=> '1.13',
+	'Locale::Maketext::GutsLoader'=> '1.13',
+	'Locale::Script'        => '2.07',
+	'MIME::Base64'          => '3.07',
+	'MIME::QuotedPrint'     => '3.07',
+	'Math::BigFloat'        => '1.60',
+	'Math::BigFloat::Trace' => '0.01',
+	'Math::BigInt'          => '1.89',
+	'Math::BigInt::Calc'    => '0.52',
+	'Math::BigInt::CalcEmu' => '0.05',
+	'Math::BigInt::Trace'   => '0.01',
+	'Math::BigRat'          => '0.22',
+	'Math::Complex'         => '1.54',
+	'Math::Trig'            => '1.18',
+	'Memoize'               => '1.01',
+	'Memoize::AnyDBM_File'  => '0.65',
+	'Memoize::Expire'       => '1.00',
+	'Memoize::ExpireFile'   => '1.01',
+	'Memoize::ExpireTest'   => '0.65',
+	'Memoize::NDBM_File'    => '0.65',
+	'Memoize::SDBM_File'    => '0.65',
+	'Memoize::Storable'     => '0.65',
+	'Module::CoreList'      => '2.17',
+	'Module::Pluggable'     => '3.8',
+	'Module::Pluggable::Object'=> '3.6',
+	'Module::Pluggable::lib::Devel::InnerPackage'=> '0.3',
+	'NDBM_File'             => '1.07',
+	'NEXT'                  => '0.61',
+	'Net::Cmd'              => '2.29',
+	'Net::Config'           => '1.11',
+	'Net::Domain'           => '2.20',
+	'Net::FTP'              => '2.77',
+	'Net::FTP::A'           => '1.18',
+	'Net::FTP::E'           => '0.01',
+	'Net::FTP::I'           => '1.12',
+	'Net::FTP::L'           => '0.01',
+	'Net::FTP::dataconn'    => '0.11',
+	'Net::NNTP'             => '2.24',
+	'Net::Netrc'            => '2.12',
+	'Net::POP3'             => '2.29',
+	'Net::Ping'             => '2.35',
+	'Net::SMTP'             => '2.31',
+	'Net::Time'             => '2.10',
+	'Net::hostent'          => '1.01',
+	'Net::netent'           => '1.00',
+	'Net::protoent'         => '1.00',
+	'Net::servent'          => '1.01',
+	'O'                     => '1.01',
+	'ODBM_File'             => '1.07',
+	'Opcode'                => '1.0601',
+	'POSIX'                 => '1.15',
+	'PerlIO'                => '1.05',
+	'PerlIO::encoding'      => '0.11',
+	'PerlIO::scalar'        => '0.06',
+	'PerlIO::via'           => '0.05',
+	'PerlIO::via::QuotedPrint'=> '0.06',
+	'Pod::Checker'          => '1.43',
+	'Pod::Find'             => '1.34',
+	'Pod::Functions'        => '1.03',
+	'Pod::Html'             => '1.09',
+	'Pod::InputObjects'     => '1.3',
+	'Pod::LaTeX'            => '0.58',
+	'Pod::Man'              => '1.37',
+	'Pod::ParseLink'        => '1.06',
+	'Pod::ParseUtils'       => '1.35',
+	'Pod::Parser'           => '1.35',
+	'Pod::Perldoc'          => '3.14',
+	'Pod::Perldoc::BaseTo'  => undef,
+	'Pod::Perldoc::GetOptsOO'=> undef,
+	'Pod::Perldoc::ToChecker'=> undef,
+	'Pod::Perldoc::ToMan'   => undef,
+	'Pod::Perldoc::ToNroff' => undef,
+	'Pod::Perldoc::ToPod'   => undef,
+	'Pod::Perldoc::ToRtf'   => undef,
+	'Pod::Perldoc::ToText'  => undef,
+	'Pod::Perldoc::ToTk'    => undef,
+	'Pod::Perldoc::ToXml'   => undef,
+	'Pod::PlainText'        => '2.02',
+	'Pod::Plainer'          => '0.01',
+	'Pod::Select'           => '1.35',
+	'Pod::Text'             => '2.21',
+	'Pod::Text::Color'      => '1.04',
+	'Pod::Text::Overstrike' => '1.1',
+	'Pod::Text::Termcap'    => '1.11',
+	'Pod::Usage'            => '1.35',
+	'SDBM_File'             => '1.06',
+	'Safe'                  => '2.16',
+	'Scalar::Util'          => '1.19',
+	'Search::Dict'          => '1.02',
+	'SelectSaver'           => '1.01',
+	'SelfLoader'            => '1.17',
+	'Shell'                 => '0.72',
+	'Socket'                => '1.81',
+	'Stdio'                 => '2.4',
+	'Storable'              => '2.19',
+	'Switch'                => '2.13',
+	'Symbol'                => '1.06',
+	'Sys::Hostname'         => '1.11',
+	'Sys::Syslog'           => '0.27',
+	'Sys::Syslog::win32::Win32'=> undef,
+	'Term::ANSIColor'       => '1.12',
+	'Term::Cap'             => '1.12',
+	'Term::Complete'        => '1.402',
+	'Term::ReadLine'        => '1.03',
+	'Test'                  => '1.25',
+	'Test::Builder'         => '0.80',
+	'Test::Builder::Module' => '0.80',
+	'Test::Builder::Tester' => '1.13',
+	'Test::Builder::Tester::Color'=> undef,
+	'Test::Harness'         => '2.64',
+	'Test::Harness::Assert' => '0.02',
+	'Test::Harness::Iterator'=> '0.02',
+	'Test::Harness::Point'  => '0.01',
+	'Test::Harness::Results'=> '0.01_01',
+	'Test::Harness::Straps' => '0.26_01',
+	'Test::Harness::Util'   => '0.01',
+	'Test::More'            => '0.80',
+	'Test::Simple'          => '0.80',
+	'Text::Abbrev'          => '1.01',
+	'Text::Balanced'        => '1.98',
+	'Text::ParseWords'      => '3.27',
+	'Text::Soundex'         => '3.03',
+	'Text::Tabs'            => '2007.1117',
+	'Text::Wrap'            => '2006.1117',
+	'Thread'                => '2.01',
+	'Thread::Queue'         => '2.11',
+	'Thread::Semaphore'     => '2.09',
+	'Thread::Signal'        => '1.00',
+	'Thread::Specific'      => '1.00',
+	'Tie::Array'            => '1.03',
+	'Tie::File'             => '0.97',
+	'Tie::Handle'           => '4.2',
+	'Tie::Hash'             => '1.03',
+	'Tie::Memoize'          => '1.1',
+	'Tie::RefHash'          => '1.38',
+	'Tie::Scalar'           => '1.01',
+	'Tie::StdHandle'        => '4.2',
+	'Tie::SubstrHash'       => '1.00',
+	'Time::HiRes'           => '1.9715',
+	'Time::Local'           => '1.1901',
+	'Time::gmtime'          => '1.03',
+	'Time::localtime'       => '1.02',
+	'Time::tm'              => '1.00',
+	'UNIVERSAL'             => '1.01',
+	'Unicode'               => '5.1.0',
+	'Unicode::Collate'      => '0.52',
+	'Unicode::Normalize'    => '1.02',
+	'Unicode::UCD'          => '0.25',
+	'User::grent'           => '1.01',
+	'User::pwent'           => '1.00',
+	'Win32'                 => '0.38',
+	'Win32API::File'        => '0.1001_01',
+	'Win32API::File::ExtUtils::Myconst2perl'=> '1',
+	'Win32CORE'             => '0.02',
+	'XS::APItest'           => '0.15',
+	'XS::Typemap'           => '0.03',
+	'XSLoader'              => '0.10',
+	'XSSymSet'              => '1.1',
+	'attributes'            => '0.09',
+	'attrs'                 => '1.02',
+	'autouse'               => '1.06',
+	'base'                  => '2.13',
+	'bigint'                => '0.23',
+	'bignum'                => '0.23',
+	'bigrat'                => '0.23',
+	'blib'                  => '1.04',
+	'bytes'                 => '1.02',
+	'charnames'             => '1.06',
+	'constant'              => '1.17',
+	'diagnostics'           => '1.16',
+	'encoding'              => '2.6_01',
+	'fields'                => '2.12',
+	'filetest'              => '1.02',
+	'if'                    => '0.05',
+	'integer'               => '1.00',
+	'less'                  => '0.01',
+	'lib'                   => '0.61',
+	'locale'                => '1.00',
+	'open'                  => '1.06',
+	'ops'                   => '1.02',
+	'overload'              => '1.06',
+	're'                    => '0.0601',
+	'sigtrap'               => '1.04',
+	'sort'                  => '1.02',
+	'strict'                => '1.03',
+	'subs'                  => '1.00',
+	'threads'               => '1.71',
+	'threads::shared'       => '1.27',
+	'utf8'                  => '1.07',
+	'vars'                  => '1.01',
+	'vmsish'                => '1.02',
+	'warnings'              => '1.05_01',
+	'warnings::register'    => '1.01',
+    },
 );
 
+# Create aliases with trailing zeros for $] use
+
+$released{'5.000'} = $released{5};
+$released{'5.010000'} = $released{5.01};
+
+$patchlevel{'5.000'} = $patchlevel{5};
+$patchlevel{'5.010000'} = $patchlevel{5.01};
+
+$version{'5.000'} = $version{5};
+$version{'5.010000'} = $version{5.01};
+
 1;
 __END__
