diff -up perl-5.34.0/ext/XS-APItest/APItest.xs.longdouble perl-5.34.0/ext/XS-APItest/APItest.xs
--- perl-5.34.0/ext/XS-APItest/APItest.xs.longdouble	2022-02-02 08:36:19.847253094 +0100
+++ perl-5.34.0/ext/XS-APItest/APItest.xs	2022-02-02 08:36:37.421344783 +0100
@@ -2334,7 +2334,7 @@ print_double(val)
 int
 have_long_double()
         CODE:
-#ifdef HAS_LONG_DOUBLE
+#if defined(HAS_LONG_DOUBLE) && defined(USE_LONG_DOUBLE)
         RETVAL = 1;
 #else
         RETVAL = 0;
