--- perl-5.8.8-RC1/utils/perlbug.PL.perlbug	2006-01-03 14:07:41.000000000 -0500
+++ perl-5.8.8-RC1/utils/perlbug.PL	2006-01-20 14:33:26.000000000 -0500
@@ -71,7 +71,7 @@
     eval 'exec $Config{perlpath} -S \$0 \${1+"\$@"}'
 	if \$running_under_some_shell;
 
-my \$config_tag1 = '$extract_version - $Config{cf_time}';
+my \$config_tag1 = '$extract_version';
 
 my \$patchlevel_date = $patchlevel_date;
 my \$patch_tags = '$patch_tags';
@@ -287,18 +287,9 @@
     # OK - send "OK" report for build on this system
     $ok = 0;
     if ($::opt_o) {
-	if ($::opt_o eq 'k' or $::opt_o eq 'kay') {
-	    my $age = time - $patchlevel_date;
-	    if ($::opt_o eq 'k' and $age > 60 * 24 * 60 * 60 ) {
-		my $date = localtime $patchlevel_date;
-		print <<"EOF";
-"perlbug -ok" and "perlbug -nok" do not report on Perl versions which
-are more than 60 days old.  This Perl version was constructed on
-$date.  If you really want to report this, use
-"perlbug -okay" or "perlbug -nokay".
-EOF
-		exit();
-	    }
+	#
+	# Red Hat modification: remove -ok / -nok refusal to report due to age
+	#
 	    # force these options
 	    unless ($::opt_n) {
 		$::opt_S = 1; # don't prompt for send
@@ -605,7 +596,7 @@
     print OUT <<EFF;
 ---
 EFF
-    print OUT "This perlbug was built using Perl $config_tag1\n",
+    print OUT "This perlbug was built using Perl $config_tag1 in the Red Hat build system.\n",
 	    "It is being executed now by  Perl $config_tag2.\n\n"
 	if $config_tag2 ne $config_tag1;
 
