diff -up perl-5.8.8/Makefile.SH.sharedlinker perl-5.8.8/Makefile.SH
--- perl-5.8.8/Makefile.SH.sharedlinker	2008-08-27 09:55:34.000000000 -0400
+++ perl-5.8.8/Makefile.SH	2008-08-27 09:56:02.000000000 -0400
@@ -80,6 +80,9 @@ true)
 		linklibperl="-L. -lperl"
 		shrpldflags="$shrpldflags -Wl,--image-base,0x57000000"
 		;;
+	linux*)
+		shrpldflags="$shrpldflags $ldflags $perllibs $cryptlib"
+		;;
 	aix*)
 		shrpldflags="-H512 -T512 -bhalt:4 -bM:SRE -bE:perl.exp"
 		case "$osvers" in
