--- perl-5.8.8/ext/threads/threads.xs.up27133_up27169	2005-10-16 13:53:32.000000000 -0400
+++ perl-5.8.8/ext/threads/threads.xs	2006-02-13 14:21:47.000000000 -0500
@@ -623,7 +623,7 @@
 	  PerlInterpreter *other_perl = thread->interp;
 	  CLONE_PARAMS clone_params;
 	  clone_params.stashes = newAV();
-	  clone_params.flags |= CLONEf_JOIN_IN;
+	  clone_params.flags = CLONEf_JOIN_IN;
 	  PL_ptr_table = ptr_table_new();
 	  current_thread = Perl_ithread_get(aTHX);
 	  Perl_ithread_set(aTHX_ thread);
--- perl-5.8.8/regcomp.c.up27133_up27169	2006-01-08 15:59:27.000000000 -0500
+++ perl-5.8.8/regcomp.c	2006-02-13 14:23:26.000000000 -0500
@@ -1810,7 +1810,7 @@
     r->subbeg = NULL;
     r->reganch = pm->op_pmflags & PMf_COMPILETIME;
     r->nparens = RExC_npar - 1;	/* set early to validate backrefs */
-
+    r->lastparen = 0;                   /* mg.c reads this.  */
     r->substrs = 0;			/* Useful during FAIL. */
     r->startp = 0;			/* Useful during FAIL. */
     r->endp = 0;			/* Useful during FAIL. */
