--- perl-5.8.5/perl.c.incorder	2005-03-08 16:11:32.722330404 -0500
+++ perl-5.8.5/perl.c	2005-03-08 16:12:04.738580672 -0500
@@ -4287,9 +4287,6 @@
     incpush(APPLLIB_EXP, TRUE, TRUE, TRUE);
 #endif
 
-#ifdef ARCHLIB_EXP
-    incpush(ARCHLIB_EXP, FALSE, FALSE, TRUE);
-#endif
 #ifdef MACOS_TRADITIONAL
     {
 	Stat_t tmpstatbuf;
@@ -4311,14 +4308,6 @@
     if (!PL_tainting)
 	incpush(":", FALSE, FALSE, TRUE);
 #else
-#ifndef PRIVLIB_EXP
-#  define PRIVLIB_EXP "/usr/local/lib/perl5:/usr/local/lib/perl"
-#endif
-#if defined(WIN32)
-    incpush(PRIVLIB_EXP, TRUE, FALSE, TRUE);
-#else
-    incpush(PRIVLIB_EXP, FALSE, FALSE, TRUE);
-#endif
 
 #ifdef SITEARCH_EXP
     /* sitearch is always relative to sitelib on Windows for
@@ -4367,6 +4356,18 @@
     incpush(PERL_OTHERLIBDIRS, TRUE, TRUE, TRUE);
 #endif
 
+#ifdef ARCHLIB_EXP
+    incpush(ARCHLIB_EXP, FALSE, FALSE, TRUE);
+#endif
+#ifndef PRIVLIB_EXP
+#  define PRIVLIB_EXP "/usr/local/lib/perl5:/usr/local/lib/perl"
+#endif
+#if defined(WIN32)
+    incpush(PRIVLIB_EXP, TRUE, FALSE, TRUE);
+#else
+    incpush(PRIVLIB_EXP, FALSE, FALSE, TRUE);
+#endif
+
     if (!PL_tainting)
 	incpush(".", FALSE, FALSE, TRUE);
 #endif /* MACOS_TRADITIONAL */
