--- perl/installperl.patch	2003-07-07 09:55:32.000000000 -0400
+++ perl/installperl	2003-07-07 09:56:03.000000000 -0400
@@ -213,7 +213,7 @@
 
 # Do some quick sanity checks.
 
-if (!$nonono && $d_dosuid && $>) { die "You must run as root to install suidperl\n"; }
+# if (!$nonono && $d_dosuid && $>) { die "You must run as root to install suidperl\n"; }
 
    $installbin		|| die "No installbin directory in config.sh\n";
 -d $installbin		|| mkpath($installbin, $verbose, 0777);
