From b4e880f3b5c8c8ba39c7c767167801a7caf81821 Mon Sep 17 00:00:00 2001
From: Tony Cook <tony@develop-help.com>
Date: Wed, 2 Jan 2019 11:49:47 +1100
Subject: [PATCH] [perl #133721] TODO test for eof with no ${^LAST_FH}
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Signed-off-by: Petr Písař <ppisar@redhat.com>
---
 t/io/tell.t | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/t/io/tell.t b/t/io/tell.t
index 7a046ca6b3..30adafbc01 100644
--- a/t/io/tell.t
+++ b/t/io/tell.t
@@ -6,7 +6,7 @@ BEGIN {
     set_up_inc('../lib');
 }
 
-plan(35);
+plan(36);
 
 $TST = 'TST';
 
@@ -187,3 +187,11 @@ seek $fh,0,0;
 is(tell, 0, "argless tell after seek \$coercible...");
 seek *$fh,0,0;
 is(tell, 0, "argless tell after seek *\$coercible...");
+
+{
+    local $TODO = "not fixed yet";
+    # [perl #133721]
+    fresh_perl_is(<<'EOI', 'ok', {}, 'eof with no ${^LAST_FH}');
+print "ok" if eof;
+EOI
+}
-- 
2.17.2

