diff -ruN perl-5.8.6-orig/lib/FindBin.pm perl-5.8.6/lib/FindBin.pm
--- perl-5.8.6-orig/lib/FindBin.pm	2003-12-27 14:52:04.000000000 +0000
+++ perl-5.8.6/lib/FindBin.pm	2005-05-16 04:15:54.000000000 +0100
@@ -179,7 +179,12 @@
       }
 
      # Get absolute paths to directories
-     $Bin     = abs_path($Bin)     if($Bin);
+     if ($Bin)
+      {
+       my $BinOld = $Bin;
+       $Bin   = abs_path($Bin);
+       defined $Bin or $Bin = File::Spec->canonpath($BinOld);
+      }
      $RealBin = abs_path($RealBin) if($RealBin);
     }
   }
