--- perl-5.8.8/Configure.BAD	2006-07-17 12:43:58.000000000 -0400
+++ perl-5.8.8/Configure	2006-07-17 12:44:04.000000000 -0400
@@ -7659,7 +7659,7 @@
 			      '') dflt="$dflt +vnocompatwarnings" ;;
 			      esac
 			      ;;	
-			linux|irix*|gnu*)	dflt='-shared' ;;
+			linux|irix*|gnu*)	dflt="-shared $optimize" ;;
 			next)  dflt='none' ;;
 			solaris) dflt='-G' ;;
 			sunos) dflt='-assert nodefinitions' ;;
