diff -up perl-5.18.0-orig/lib/perl5db.pl.orig perl-5.18.0-orig/lib/perl5db.pl
diff -up perl-5.18.0-orig/lib/perl5db.t.orig perl-5.18.0-orig/lib/perl5db.t
--- perl-5.18.0-orig/lib/perl5db.t.orig	2013-06-11 16:56:29.291375625 +0200
+++ perl-5.18.0-orig/lib/perl5db.t	2013-06-12 13:51:32.268233583 +0200
@@ -2286,6 +2286,7 @@ sub _calc_trace_wrapper
 
 # Test the o AutoTrace command
 {
+    local $ENV{PERL_RL} = "o=0";
     my $wrapper = DebugWrap->new(
         {
             cmds =>
